test/model/message_list_test.dart

```diff
@@ -162,5 +162,107 @@ void main() async {
     test('rendering-only update does not change timestamp (for old server versions)', () async {
       await checkRenderingOnly(legacy: true);
     });
+
+    group('ReactionEvent handling', () {
+      ReactionEvent mkEvent(Reaction reaction, ReactionOp op, int messageId) {
+        return ReactionEvent(
+          id: 1,
+          op: op,
+          emojiName: reaction.emojiName,
+          emojiCode: reaction.emojiCode,
+          reactionType: reaction.reactionType,
+          userId: reaction.userId,
+          messageId: messageId,
+        );
+      }
+
+      test('add reaction', () async {
+        final originalMessage = eg.streamMessage(id: 243, stream: stream, reactions: []);
+        final messageList = await messageListViewWithMessages([originalMessage], stream, narrow);
+
+        final message = messageList.messages.single;
+        check(message).reactions.not(it()..jsonEquals([eg.unicodeEmojiReaction]));
+
+        bool listenersNotified = false;
+        messageList.addListener(() { listenersNotified = true; });
+
+        messageList.maybeUpdateMessageReactions(
+          mkEvent(eg.unicodeEmojiReaction, ReactionOp.add, originalMessage.id));
+
+        check(listenersNotified).isTrue();
+        check(messageList.messages.single)
+          ..identicalTo(message)
+          ..reactions.jsonEquals([eg.unicodeEmojiReaction]);
+      });
+
+      test('add reaction; message is not in list', () async {
+        final someMessage = eg.streamMessage(id: 1, reactions: []);
+        final messageList = await messageListViewWithMessages([someMessage], stream, narrow);
+        check(messageList.messages.single).reactions.jsonEquals([]);
+
+        bool listenersNotified = false;
+        messageList.addListener(() { listenersNotified = true; });
+
+        messageList.maybeUpdateMessageReactions(
+          mkEvent(eg.unicodeEmojiReaction, ReactionOp.add, 1000));
+
+        check(listenersNotified).isFalse();
+        check(messageList.messages.single).reactions.jsonEquals([]);
+      });
+
+      test('remove reaction', () async {
+        final eventReaction = Reaction(reactionType: ReactionType.unicodeEmoji,
+          emojiName: 'wave',                  emojiCode: '1f44b', userId: 1);
+
+        // Same emoji, different user. Not to be removed.
+        final reaction2 = Reaction.fromJson(eventReaction.toJson()..update('user_id', (_) => 2));
```

Comment:
```suggestion
        final reaction2 = Reaction.fromJson(eventReaction.toJson()
          ..update('user_id', (_) => 2));
```

(to make the parallelism clear with the others below)